### ISSUE TITLE
Questions regarding the algorithm

### ISSUE THREAD

**Issue by varun-intel** [caslab-vt/SARNet#2]
I have a few question about your training algorithm:
1. How are shared policy parameters updated? From my understanding, it seems you are updating them once in each agent that uses the shared params. Is this correct?
2. How many rounds of communication are you performing per environment time step? I see only one round, as opposed to 2 or more rounds in TarMAC.
3. In the policy optimization step, how do you compute the fresh actions that will be passed in to the q function? It seems that you are first taking the actions from the replay buffer and only replacing the action for the agent being optimized; is this correct? Also, when you compute this action, are you generating new messages in the forward pass through all the agents? My concern here is that there will be a mismatch between the fresh messages and the stale actions of the other agents.

Thanks for your help.

**Comment by murtazarang** [comment]
1. The network of each agent is created as a separate instance (partially shared) and updated independently at each update step.
2. For the scope of the paper we perform a single step of communication round. However, we can also perform multiple communication rounds, similar to CommNet or TarMAC.

3a. We store the hidden states of the q values for each environment step in the buffer. This initial hidden states are then used to generate a trajectory to perform BPTT updates. Consequently, you could also use a initial hidden state of the Q as a zero vector for generating the BPTT updates. 
3b. Yes, that's correct.
3c. Similar to the ones of how we generate the rollout new hidden states for the Q value for BPTT updates, we perform the same rollout for the messages and hidden states of the actor for actor updates. 

Hope that answers your query.

**Comment by varun-intel** [comment]
Thanks; yes, that answers my questions. Could you clarify a couple small points?
1) Suppose agents 1 and 2 share parameters, and we are currently optimizing agent 1. Do gradients flow back through the messages sent by agent 2?
2) The gating operation used in the IC3Net function appears to be a no-op. It does a softmax over a 1-length dimension, which will always produce 1. Is this supposed to happen?

**Comment by murtazarang** [comment]
1. Each agent maintains a copy of the parameters used to generate the communication messages of other agents (key/value). During optimization of agent 1, the parameters used to generate the key/value of agent 2 are also updated, which is then reflected across other agents (2, 3, 4).
2. Thank you for noting a bug in the code on Github. The softmax op is a mistake in the code, I'll need to verify the results, and get back to you in about a week or two.

**Comment by murtazarang** [comment]
Each agent maintains a copy of the parameters used to generate the communication messages of other agents (key/value). During optimization of agent 1, the parameters used to generate the key/value of agent 2 are also updated, which is then reflected across other agents (2, 3, 4 ...).


Thank you for noting a bug in the code on Github. The softmax op is a mistake in the code, I'll need to look into the issue, and get back to you in about a week or two.

**Comment by varun-intel** [comment]
Thanks again. One more point: could you please explain the meaning of the 'soft categorical' policy distribution, which seems to be used in the mpe environments? As far as I can tell, the output of the policy net is a distribution over discrete actions. However, I don't see an action being sampled from the distribution, and the environment seems to use all the entries in the vector. Why is the soft categorical used as opposed to just using a linear layer, as is standard in td3?

**Comment by murtazarang** [comment]
Sorry about the delayed response.

To answer your previous query on the softmax operation always setting the communication as 1 in IC3Net, please refer the to Section 4.2 in the IC3Net paper. The authors set the gating action to always "1" in fully cooperative settings. For predator-prey tasks, we also modeled the communication networks, separately for predators and preys, instead of a singular network, to avoid information leak between agents. However, I will add the IC3Net gating action sampling in the future to the repo, to keep it more consistent.

For SoftCategoricalPD, the distribution is used to add exploration noise directly to the outputs of the network, which is an implementation of Gumbel-Softmax. I found the method to perform more consistently than using a direct linear layer as the output, when working within the multi-agent particle environment.
